storage/storage_image.go

```diff
@@ -93,309 +90,655 @@ func newImageSource(imageRef storageReference) (*storageImageSource, error) {
+	mb, err := m.Serialize()
+	if err != nil {
+		return ""
+	}
+	switch manifest.GuessMIMEType(mb) {
+	case manifest.DockerV2Schema2MediaType, imgspecv1.MediaTypeImageManifest:
+		// For Schema2 and OCI1(?), the ID is just the hex part of the digest of the config blob.
+		logrus.Debugf("trivial image ID for configured blob")
+		configInfo := m.ConfigInfo()
+		if configInfo.Digest.Validate() == nil {
+			return configInfo.Digest.Hex()
+		}
+		return ""
+	case manifest.DockerV2Schema1MediaType, manifest.DockerV2Schema1SignedMediaType:
+		// Convert the schema 1 compat info into a schema 2 config, constructing some of the fields
+		// that aren't directly comparable using info from the manifest.
+		logrus.Debugf("computing image ID using compat data")
+		s1, ok := m.(*manifest.Schema1)
+		if !ok {
+			logrus.Debugf("schema type was guessed wrong?")
+			return ""
+		}
+		if len(s1.History) == 0 {
+			logrus.Debugf("image has no layers")
+			return ""
+		}
+		s2 := struct {
+			manifest.Schema2Image
+			ID        string `json:"id,omitempty"`
+			Parent    string `json:"parent,omitempty"`
+			ParentID  string `json:"parent_id,omitempty"`
+			LayerID   string `json:"layer_id,omitempty"`
+			ThrowAway bool   `json:"throwaway,omitempty"`
+			Size      int64  `json:",omitempty"`
+		}{}
+		config := []byte(s1.History[0].V1Compatibility)
+		if json.Unmarshal(config, &s2) != nil {
+			logrus.Debugf("error decoding configuration")
+			return ""
+		}
+		// Images created with versions prior to 1.8.3 require us to rebuild the object.
+		if s2.DockerVersion != "" && versions.LessThan(s2.DockerVersion, "1.8.3") {
+			err = json.Unmarshal(config, &s2)
+			if err != nil {
+				logrus.Infof("error decoding compat image config %s: %v", string(config), err)
+				return ""
+			}
+			config, err = json.Marshal(&s2)
+			if err != nil {
+				logrus.Infof("error re-encoding compat image config %#v: %v", s2, err)
+				return ""
+			}
+		}
+		// Build the history.
+		for _, h := range s1.History {
+			compat := manifest.Schema1V1Compatibility{}
+			if json.Unmarshal([]byte(h.V1Compatibility), &compat) != nil {
+				logrus.Debugf("error decoding history information")
+				return ""
+			}
+			hitem := manifest.Schema2History{
+				Created:    compat.Created,
+				CreatedBy:  strings.Join(compat.ContainerConfig.Cmd, " "),
+				Comment:    compat.Comment,
+				EmptyLayer: compat.ThrowAway,
+			}
+			s2.History = append([]manifest.Schema2History{hitem}, s2.History...)
+		}
+		// Build the rootfs information.  We need the decompressed sums that we've been
+		// calculating to fill in the DiffIDs.
+		s2.RootFS = &manifest.Schema2RootFS{
+			Type: "layers",
+		}
+		for _, fslayer := range s1.FSLayers {
+			blobSum := fslayer.BlobSum
+			diffID, ok := s.blobDiffIDs[blobSum]
+			if !ok {
+				logrus.Infof("error looking up diffID for blob %q", string(blobSum))
+				return ""
+			}
+			s2.RootFS.DiffIDs = append([]digest.Digest{diffID}, s2.RootFS.DiffIDs...)
```

Comment:
AFAICT layers marked as `ThrowAway` in `history` are not added to `DiffIDs` in docker/docker.

Reply:
Yup, fixing this.